BugFix: improve handling of TConsole::mSystemMessageBgColor

SlySven opened the pull request:
This `QColor` is initilised to the default value of `TConsole::mBgColor` which, since #4570, is now the transparent colour. Unfortunately this is bad because when text printed on the main console with: `(void) TConsole::printSystemMessage(const QString)` the red foreground and the transparent background get swapped over and thus is rendered as a solid block of red when it is selected.

This PR attempts to solve this by ensuring that when the `TConsole::mBgColor` member is set to a different colour that the same colour is also stored for use as `TConsole::mSystemMessageBgColor`.

The issue of how things are coloured does seem to be made more complicated by the use of `QPalette` for some things for which it does not seem to be worthwhile; therefore in this PR I also remove some uses of the Qt palette system in places where it does not seem to be doing anything useful.

I have also reviewed how things might be affected by setting an image as a background image for the main console but it seems that the changes here do not prevent the display of an image where there is NO text shown on screen.

This might close #4635.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: improve handling of TConsole::mSystemMessageBgColor

This `QColor` is initilised to the default value of `TConsole::mBgColor`
which, since #4570, is now the transparent color. Unfortunely this is bad
because when text printed on the main console with:
`(void) TConsole::printSystemMessage(const QString)`
the red foreground and the transparent background get swapped over and thus
is rendered as a solid block of red.

This PR attempts to solve this by ensuring that whne the
`TConsole::mBgColor` member is set to a different color that the same
colour is also stored for use as `TConsole::mSystemMessageBgColor`.

The issue of how things are colored does seem to be made more complicated
by the use of `QPalette` for some things for which it does not seem to be
worthwhile; therefore in this PR I also remove some uses of the Qt palette
system in places where it does not seem to be doing anything useful.

I have also reviewed how things might be affected by setting an image as
a background image for the main console but it seems that the changes here
do not prevent the display of an image where there is NO text shown on
screen.

This might close #4635.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TConsole.cpp:
```diff
@@ -453,13 +437,13 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
     mpBufferSearchBox->setFont(mpHost->mCommandLineFont);
     mpBufferSearchBox->setFocusPolicy(Qt::ClickFocus);
     mpBufferSearchBox->setPlaceholderText("Search ...");
-    QPalette __pal;
```
:boom: An identifier ***containing*** a double underscore is, apparently, **forbidden** by C++ identifier rules. https://stackoverflow.com/a/228797/4805858 ...

SlySven commented on src/TConsole.cpp:
```diff
@@ -1648,21 +1600,20 @@ void TConsole::setDisplayAttributes(const TChar::AttributeFlags attributes, cons
 void TConsole::setFgColor(int r, int g, int b)
 {
     setFgColor(QColor(r, g, b));
-    mUpperPane->forceUpdate();
```
These cause an unneeded redraw as the method called on the previous line already does it.

SlySven commented on src/TConsole.cpp:
```diff
@@ -1648,21 +1600,20 @@ void TConsole::setDisplayAttributes(const TChar::AttributeFlags attributes, cons
 void TConsole::setBgColor(int r, int g, int b, int a)
 {
     setBgColor(QColor(r, g, b, a));
-    mUpperPane->forceUpdate();
```
These cause an unneeded redraw as the method called on the previous line already does it.

SlySven commented on src/TConsole.cpp:
```diff
@@ -152,19 +152,10 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
-    splitterPalette.setColor(QPalette::Window, QColor(Qt::green)); //,255) );
-    splitterPalette.setColor(QPalette::Base, QColor(255, 0, 0, 255));
-    splitterPalette.setColor(QPalette::Window, QColor(Qt::white));
-    //setPalette( mainPalette );
```
The `splitterPalette` has already become disused here...

SlySven commented on src/TConsole.cpp:
```diff
@@ -192,13 +183,6 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
-    topbarPalette.setColor(QPalette::Highlight, QColor(55, 55, 255));
-    topbarPalette.setColor(QPalette::Window, QColor(0, 255, 0, 255));
-    topbarPalette.setColor(QPalette::Base, QColor(0, 255, 0, 255));
-    //mpTopToolBar->setPalette(topbarPalette);
```
The `topbarPalette` has already become disused here...

SlySven commented on src/TConsole.cpp:
```diff
@@ -930,27 +912,13 @@ void TConsole::changeColors()
-        palette.setColor(QPalette::Text, mFgColor);
-        palette.setColor(QPalette::Highlight, QColor(55, 55, 255));
-        palette.setColor(QPalette::Base, QColor(Qt::black));
-        mUpperPane->setPalette(palette);
```
All the drawing and highlighting is performed by us - so this and the similar stuff further down is pointless.

vadi2 reviewed (commented):
This seems like a lot of changes for a relatively small fix...and won't this work only for system messages? If other messages have a transparent background, they'll have the same problem?

I would approach it by simply taking care of transparent background in the selection code, since that is where the problem actually is.

SlySven commented:
This code In use - I have revised the background colour in the preferences by making it a darker grey, between the first replay start/stop and the second one, and (though not shown) selecting the text does use swap the (non-transparent) colours over:

![image](https://user-images.githubusercontent.com/6163092/104845110-a680db00-58cb-11eb-9812-7ae4f33d4a0a.png)

![image](https://user-images.githubusercontent.com/6163092/104845204-05deeb00-58cc-11eb-92a9-07dd4380dd42.png)

SlySven commented:
Only just spotted your reply after posting the above.

> I would approach it by simply taking care of transparent background in the selection code.

😕 How?

It seems that using transparency was just an initialisation issue - which this code now avoids by making sure that the background colour is *always* used for the system message whereas it was previously being set on construction and not being updated - even to what was subsequently set.

~If w~*W*e *do (unfortunately)* provide a means for the end-user to use a transparent background for the console - ~and it isn't immediately obvious to me whether it is,~ then what possible background (or foreground if the text *is* reversed or {exclusively} selected) are we supposed to use?

SlySven commented:
I note that the method in the preferences for setting the background colour of the main console does NOT provide access to an `alpha` setting...

SlySven commented:
Any solution at the point of painting the text is going to be impacted by #4645 - to provide the reversed effect would require drawing the text glyphs in a manner where the pixels in the glyph are the background and the remainder are in red for the system message or the foreground colour for other text - however that will increase the complexity of that code as I am not sure of a simple way in order to *cut-out* the glyph from the background colour.

vadi2 commented:
So do we really need to support the reversed effect?

SlySven commented:
Well it was introduced two years ago #1840 and thus made it into the Lua API. I guess I will have to revisit #4645 (after it has been approved if possible, 🥺) to see if I can improve the way it works (the first pass through each line that draws the background will need enhancement to - if the *foreground* has transparency - paint the glyph as part of the background)...

vadi2 commented:
What's left outstanding for this PR?

SlySven commented:
> What's left outstanding for this PR?

Nothing, I think it is good enough now. It ensures that the system message background color tracks whatever background colour is actually in use for the main `TConsole`, rather than being left in an initialisatation value of being transparent. The PR also cleans out some ineffective (I think) `QPalette` cruft.

The separate matter of what to do if a transparent background is chosen is a bit relevant but it is it the subject of a separate issue IMHO.

Edru2 commented:
Why are there no files to test the PR?
I tested it and it seems to break changing of command line colour in settings.
~~Didn't check which part exactly causes this.~~
edit: see below

Edru2 commented on src/TConsole.cpp:
```diff
@@ -959,13 +927,6 @@ void TConsole::changeColors()
         if (mpCommandLine) {
             auto styleSheet = mpCommandLine->styleSheet();
             mpCommandLine->setStyleSheet(QString());
-            QPalette pal;
-            pal.setColor(QPalette::Text, mpHost->mCommandLineFgColor); //QColor(0,0,192));
-            pal.setColor(QPalette::Highlight, QColor(0, 0, 192));
-            pal.setColor(QPalette::HighlightedText, QColor(Qt::white));
-            pal.setColor(QPalette::Base, mpHost->mCommandLineBgColor); //QColor(255,255,225));
-            mpCommandLine->setPalette(pal);
-            mpCommandLine->mRegularPalette = pal;
```
okay after testing. It is this part which messes up with changing command line background and textcolor through settings.

SlySven replied:
:man_facepalming:

Ah, yeah I might have gone too far there...

... let me adjust things a bit. :hammer:

SlySven replied:
Done in https://github.com/Mudlet/Mudlet/pull/4650/commits/bb2da9d4943c5eee6086e1045815127005697a7b .

Edru2 replied:
Seems to be fixed. :+1:

SlySven committed:
Merge branch 'development' into BugFix_improveHandlingOf_TConsole__mSystemMessageBgColor

Bring in upstream changes

SlySven committed:
BugFix: part revert previous commit and correct spelling

The modifications made to `(void) TConsole::changeColors()` to take away
the manipulation of the command line palette should not have been done!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
